substrate/client/network/src/litep2p/shim/request_response/tests.rs

```diff
@@ -321,3 +322,322 @@ async fn feedback_works() {
+	}
+
+	// sender: oneshot::Sender<Result<(Vec<u8>, ProtocolName), RequestFailure>>,
+	match result_rx.await {
+		Ok(Ok((response, protocol))) => {
+			assert_eq!(response, vec![1, 3, 3, 8]);
+			assert_eq!(protocol, ProtocolName::from("/protocol/1"));
+		},
+		event => panic!("invalid event: {event:?}"),
+	}
+}
+
+#[tokio::test]
+async fn fallback_request_old_peer_sends() {
+	// peer1 supports new, binary-incompatible version of the protocol
+	let (config1, handle1) = ConfigBuilder::new(litep2p::ProtocolName::from("/protocol/2"))
+		.with_fallback_names(vec![litep2p::ProtocolName::from("/protocol/1")])
+		.with_max_size(1024)
+		.build();
+
+	let mut litep2p1 = Litep2p::new(
+		Litep2pConfigBuilder::new()
+			.with_request_response_protocol(config1)
+			.with_tcp(TcpConfig {
+				listen_addresses: vec![
+					"/ip4/0.0.0.0/tcp/0".parse().unwrap(),
+					"/ip6/::/tcp/0".parse().unwrap(),
+				],
+				..Default::default()
+			})
+			.build(),
+	)
+	.unwrap();
+
+	// peer2 supports only the old version of the protocol
+	let (config2, handle2) = ConfigBuilder::new(litep2p::ProtocolName::from("/protocol/1"))
+		.with_max_size(1024)
+		.build();
+
+	let mut litep2p2 = Litep2p::new(
+		Litep2pConfigBuilder::new()
+			.with_request_response_protocol(config2)
+			.with_tcp(TcpConfig {
+				listen_addresses: vec![
+					"/ip4/0.0.0.0/tcp/0".parse().unwrap(),
+					"/ip6/::/tcp/0".parse().unwrap(),
+				],
+				..Default::default()
+			})
+			.build(),
+	)
+	.unwrap();
+
+	let peer1 = *litep2p1.local_peer_id();
+	let peer2 = *litep2p2.local_peer_id();
+
+	connect_peers(&mut litep2p1, &mut litep2p2).await;
+
+	let (tx1, rx1) = async_channel::bounded(4);
+	let (protocol1, _outbound_tx1) = RequestResponseProtocol::new(
+		ProtocolName::from("/protocol/2"),
+		handle1,
+		Arc::new(peerstore_handle_test()),
+		Some(tx1),
+		None,
+	);
+
+	let (tx2, _rx2) = async_channel::bounded(4);
+	let (protocol2, outbound_tx2) = RequestResponseProtocol::new(
+		ProtocolName::from("/protocol/1"),
+		handle2,
+		Arc::new(peerstore_handle_test()),
+		Some(tx2),
+		None,
+	);
+
+	tokio::spawn(protocol1.run());
+	tokio::spawn(protocol2.run());
+	tokio::spawn(async move { while let Some(_) = litep2p1.next_event().await {} });
+	tokio::spawn(async move { while let Some(_) = litep2p2.next_event().await {} });
+
+	let (result_tx, result_rx) = oneshot::channel();
+	outbound_tx2
+		.unbounded_send(OutboundRequest {
+			peer: peer1.into(),
+			request: vec![1, 2, 3, 4],
+			sender: result_tx,
+			fallback_request: None,
+			dial_behavior: IfDisconnected::ImmediateError,
+		})
+		.unwrap();
+
+	// FIXME: because `/protocol/1` is a fallback of `/protocol/2`, requests over both protocols
+	// are received over the same channel which in the high-level API is tied to one single protocol
+	// and the actual negotiated protocol is reported to the request handler
+	//
+	// this means that request handler doesn't know the protocol of the request and may decode the
+	// request incorrectly
```

Comment:
Do you plan to implement this in this PR? Otherwise, IMO, the PR is ready to be approved.

Reply:
Yes I'll have to fix it because otherwise it would block the other PR that's relying on this feature but I'm looking into Alin's proposal.